### PR TITLE
implements `get_next_block_best_parents` + `get_largest_stakers` methods

### LINKED ISSUE #3419: Implement `get_next_block_best_parents` method in gRPC API
This method will be a part of the new gRPC API as it's needed to be used primarily in newer Massa architecture.

Related to #3406 #2895 #2903

### PR DESCRIPTION
**Pull request by modship**
* [x] document all added functions
* [X] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [X] add logs allowing easy debugging in case the changes caused problems
* [X] if the API has changed, update the API specification

closes #3419 #3485

### CODE REVIEW

**Comment by modship** [comment]
@aoudiamoncef 

I should update openrpc.json file manually ?

**Comment by aoudiamoncef** [comment]
> @aoudiamoncef
> 
> I should update openrpc.json file manually ?

Yes

File: massa-node/base_config/openrpc.json [hunk 1]
```diff
@@ -1306,6 +1306,34 @@
             "name": "unsubscribe_new_operations",
             "summary": "Unsubscribe from new received operations",
             "description": "Unsubscribe from new received operations."
+        },
+        {
+            "tags": [
+                {
+                    "name": "api",
+                    "description": "Massa api V2"
+                }
```
Comment (aoudiamoncef) [inline massa-node/base_config/openrpc.json:1315]: Please, add experimental tag here

Comment (modship) [reply massa-node/base_config/openrpc.json:1319]: Done

File: massa-api/src/api.rs [hunk 2]
```diff
@@ -51,6 +55,16 @@ impl MassaApiServer for API<ApiV2> {
         Ok(self.0.version)
     }
 
+    fn get_best_parents(
+        &self,
+        page_request: Option<PageRequest>,
```
Comment (aoudiamoncef) [inline massa-api/src/api.rs:60]: In API V2, we should anticipate and avoid breaking changes, we should create a struct ApiRequest to wrap all existing and coming parameters

Comment (modship) [reply massa-api/src/api.rs]: Done

Comment (modship) [reply massa-api/src/api.rs]: I forgot to update openrpc.json, i'll do it

Comment (modship) [reply massa-api/src/api.rs]: done

File: massa-api/src/api_request.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+use massa_api_exports::page::PageRequest;
+use serde::{Deserialize, Serialize};
+
+/// Wrap request params into struct for ApiV2 method
+#[derive(Deserialize, Serialize)]
+pub struct ApiRequest {
```
Comment (aoudiamoncef) [inline massa-api/src/api_request.rs:6]: this should be moved to `massa-api-exports` as it's used externally.

**Review by aoudiamoncef — CHANGES_REQUESTED** [review]
@modship, 

- May be we should rename this method to `get_next_block_best_parents` or `next_block_best_parents` ? 

- I was thinking about flatten or not ApiRequest struct, in the future, we could add `sorting`, `filter` and `args` for example. IMHO the actual form is good.

**Review by aoudiamoncef — APPROVED** [review]
Great work @modship 👍🏻

File: massa-api-exports/src/api_request.rs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+use crate::page::PageRequest;
+use serde::{Deserialize, Serialize};
+
+/// Wrap request params into struct for ApiV2 method
+#[derive(Deserialize, Serialize)]
+pub struct ApiRequest {
```
Comment (aoudiamoncef) [inline massa-api-exports/src/api_request.rs:6]: As it's common and small, we could add it directly to `lib.rs` but may be in the future, we'll create `types.rs`or `common.rs`to keep `lib.rs`clean

**Review by aoudiamoncef — CHANGES_REQUESTED** [review]

File: massa-api-exports/src/api_request.rs (hunk 4 above)
Comment (modship) [reply massa-api-exports/src/api_request.rs]: Done, 

If u want i can squash my commit on this branch

File: massa-node/base_config/openrpc.json [hunk 5]
```diff
@@ -1306,6 +1306,39 @@
...
+                "schema": {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/components/schemas/Parent"
+                    }
+                },
+                "name": "best parents"
+            },
+            "name": "get_next_block_best_parents",
+            "summary": "Get best parents",
```
Comment (aoudiamoncef) [inline massa-node/base_config/openrpc.json:1340]: Net next block best parents

File: massa-node/base_config/openrpc.json [hunk 6]
```diff
@@ -1306,6 +1306,39 @@
...
+                }
+            ],
+            "result": {
+                "schema": {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/components/schemas/Parent"
+                    }
+                },
+                "name": "best parents"
```
Comment (aoudiamoncef) [inline massa-node/base_config/openrpc.json:1337]: Next_block best parents

**Review by aoudiamoncef — CHANGES_REQUESTED** [review]

**Comment by aoudiamoncef** [comment]
Hi @modship @Leo-Besancon
 I continued the work on this PR by removing pagination from `get_next_block_best_parents` and reusing the code in `get_largest_stakers`
 
 In this PR, we are adding an evolutive way to do request parameters and pagination. We'll try to add this to testnet19 and collect feedbacks from our consumers. 
 
I'll update the Postman collection with this new experimental methods.

File: massa-api-exports/src/lib.rs [hunk 7]
```diff
@@ -80,3 +81,10 @@ pub enum ListType {
     /// contains allowed entry
     Whitelist,
 }
+
+/// Wrap request params into struct for ApiV2 method
+#[derive(Deserialize, Serialize)]
```
Comment (Leo-Besancon) [inline massa-api-exports/src/lib.rs:86]: Why do we wrap the PageRequest here? To have a new struct we can freely make changes to for APIv2 without breaking the APIv1?

**Review by Leo-Besancon — COMMENTED** [review]
This APIv2 will be great 👌

File: massa-api-exports/src/lib.rs (hunk 7 above)
Comment (aoudiamoncef) [reply massa-api-exports/src/lib.rs:86]: We should have an evolutive struct in APIV2, this is a good practice in API design, for example, if we would add Sort struct without breaking our existing API, we'll just add to ApiRequest. In addition, JsonRPC takes an array of objects as query parameters:

BAD
```shell
curl --location --request POST 'http://localhost:33036' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get_largest_stakers",
    "params": [ {{"sort_by_id": "ASC", "offset": 0, "limit": 2 }]
}'
```
We should play this flatten object and bind it manually to our structs: Sort PageRequest. if we have same parameters names between different args, here we are ...

BETTER

```shell
curl --location --request POST 'http://localhost:33036' \
--header 'Content-Type: application/json' \
--data-raw '{
    "jsonrpc": "2.0",
    "id": 1,
    "method": "get_largest_stakers",
    "params": [ {"sort": {"id": "ASC"}, "page_request": {"offset": 0, "limit": 2 }}]
}'
```
Nothing to do, ApiRequest will hold all for us.

Comment (Leo-Besancon) [reply massa-api-exports/src/lib.rs:86]: Understood, it's what I thought ;). Nice addition then 👍

**Comment by aoudiamoncef** [comment]
@Leo-Besancon,

I added a default offset limit on `get_largest_stakers` to return the first 50 stakers. 
.
IMHO, this is fine for an experimental feature, we'll collect the feedbacks in testnet19 and improve things.

**Review by Leo-Besancon — APPROVED** [review]
Looks fine to me 👍

**Review by aoudiamoncef — APPROVED** [review]

**Comment by AurelienFT** [comment]
@modship can you rebase this on the latest version of testnet_20 ? the CI has been fixed and you will be able to merge it.

**Pull request merged**